### ISSUE TITLE
Head-wise RMSNorm applied before or after the gating?

### ISSUE THREAD

**Issue by alxndrTL** [qiuzh20/gated_attention#1]
Hello,
Thank you very much for open-sourcing this work!

In the paper, it is stated that you applied a norm for each head independently, before the concatenation:

_Inspired by prior works that utilize group norm for the SDPA output (Sun et al., 2023; Ye et al., 2024),
with the same setting in Sec. 3.2.1, we apply RMSNorm (Zhang & Sennrich, 2019) independently to the
output of each attention head before concatenation_

I wonder where exactly this norm is placed relative to the gating ? I couldn't find it in the code.

Thank you,
Best

**Comment by qiuzh20** [comment]
Thank you for your interest!

In fact, we primarily provided the code for using gating. More ablation study variants were not included in the cleaned-up code.

Regarding RMSNorm, we did not use RMSNorm in conjunction with gating. We observed that training was very stable after adding gating, and the sparsity of different heads varied significantly, so we didn't introduce group norm in that case.

We only added group norm to standard softmax attention as an analysis to add non-linearity.

**Comment by alxndrTL** [comment]
Ok thank you for your detailed and fast response!
